### PR TITLE
[14.0][MIG] stock_valuation_layer_usage and stock_account_product_run_fifo_hook

### PR DESCRIPTION
**Pull request by ChrisOForgeFlow**
@ForgeFlow

Introduced new field to check original layer on outcome layers
Removed reservation case, because on core don't get any traceability data to assign candidate layer, only get by order and quantity available(field remaining_qty)

### CODE REVIEW

**Review by AaronHForgeFlow — APPROVED** [review]
Functional review. Working fine. This is not WIP anymore, is it?

File: stock_account_product_run_fifo_hook/hooks.py [hunk 1]
```diff
@@ -18,12 +18,7 @@ def _run_fifo_new(self, quantity, company):
         qty_to_take_on_candidates = quantity
         # START HOOK Search Candidates
         candidates_domain = self._get_candidates_domain(company)
-        candidates = (
-            self.env["stock.valuation.layer"]
-            .sudo()
-            .with_context(active_test=False)
```
Comment (MiquelRForgeFlow) [inline stock_account_product_run_fifo_hook/hooks.py]: why?

File: stock_valuation_layer_usage/__manifest__.py [hunk 2]
```diff
@@ -3,15 +3,15 @@
...
+    "version": "14.0.1.0.0",
     "category": "Warehouse Management",
     "development_status": "Alpha",
     "license": "AGPL-3",
     "summary": "Trace where has the stock valuation been used in, including "
     "the quantities taken.",
     "author": "ForgeFlow,Odoo Community Association (OCA)",
     "website": "https://github.com/OCA/stock-logistics-workflow",
-    "depends": ["stock_account_product_run_fifo_hook"],
+    "depends": ["sale", "stock_account_product_run_fifo_hook"],
```
Comment (MiquelRForgeFlow) [inline stock_valuation_layer_usage/__manifest__.py:14]: why?

File: stock_valuation_layer_usage/models/stock_valuation_layer.py [hunk 3]
```diff
@@ -29,11 +29,13 @@ class StockValuationLayer(models.Model):
...
-                    ("stock_move_id", "=", rec.stock_move_id.id),
-                ]
-            )
+            domain = [
+                ("stock_move_id", "=", rec.stock_move_id.id),
+            ]
+            lots = rec.stock_move_id.mapped("move_line_ids.lot_id").ids
+            if lots:
+                domain += [("stock_move_id.move_line_ids.lot_id", "in", lots)]
+            incoming_usages = self.env["stock.valuation.layer.usage"].search(domain)
```
Comment (rousseldenis) [inline stock_valuation_layer_usage/models/stock_valuation_layer.py]: @ChrisOForgeFlow Isn't there a mean to avoid this search in the loop ?

**Review by JordiBForgeFlow — APPROVED** [review]

**Review by AaronHForgeFlow — APPROVED** [review]

**Review by MiquelRForgeFlow — APPROVED** [review]
@ChrisOForgeFlow please, squash commits in a beautiful and meaningful way :)

**Comment by MiquelRForgeFlow** [comment]
@rousseldenis @JordiBForgeFlow the PR is ready

**Review by sysadminmatmoz — APPROVED** [review]
It works for me... tested extensively in a big database with +100000s stock moves, abusing and simulating bad behaviour. Used a similar hook (identical in functionality) in production version 12 for 3 years.

**Comment by AaronHForgeFlow** [comment]
ready to merge IMHO

**Pull request merged**